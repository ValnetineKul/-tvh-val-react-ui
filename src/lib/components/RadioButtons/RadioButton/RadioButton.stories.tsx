import { useArgs } from '@storybook/client-api';
import type { ComponentProps } from 'react';
import React from 'react';
import type { StoryTemplate } from '../../../story-utils';
import { createMeta } from '../../../story-utils';
import RadioButton from './RadioButton';

type Props = ComponentProps<typeof RadioButton>;
type Args = Omit<Props, 'onChange'>;

const Template = ((args: Args) => {
  const [{ checked }, updateArgs] = useArgs();

  return (
    <RadioButton
      {...args}
      checked={checked}
      onChange={(...params) => {
        updateArgs({ checked: params[1] });
      }}
    />
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  label: 'RadioButton label',
  value: 'RadioButton value',
  name: 'radio-button',
  checked: false,
  disabled: false,
};

Primitive.storyName = 'RadioButton';

export default createMeta({
  component: RadioButton,
  title: 'Components/RadioButtons/RadioButton',
  argTypes: {
    onChange: {
      action: { argTypesRegex: '^on.*' },
      table: {
        type: {
          detail:
            'Callback fired when the state is changed: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void.',
        },
      },
    },
    value: {
      table: {
        type: {
          detail: 'Value of the selected radio button.',
        },
      },
    },
  },
});
