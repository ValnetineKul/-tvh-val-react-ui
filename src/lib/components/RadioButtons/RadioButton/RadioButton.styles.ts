import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles<void, 'disabled' | 'label'>({ name: 'RadioButton' })((theme, _, classes) => ({
  root: {
    alignItems: 'flex-start',
    marginLeft: theme.layoutSpacing['spacing/000'],
    marginRight: theme.layoutSpacing['spacing/000'],
    [`&.${classes.disabled} .${classes.label}`]: {
      color: theme.color['text/onSurfaceCurrent/disabled'],
    },
  },
  radio: {
    padding: theme.layoutSpacing['spacing/000'],
    margin: theme.layoutSpacing['spacing/100'],
    backgroundColor: theme.color['bg/surface100/default'],
    '&:hover': {
      backgroundColor: theme.color['bg/surface100/default'],
    },
    '&:focus-within': {
      boxShadow: `0px 0px 0px 4px ${theme.color['border/onSurfaceCurrent/focus']}`,
    },
  },
  disabled: {},
  label: {
    marginLeft: theme.layoutSpacing['spacing/300'],
  },
  icon: {
    width: `calc( ${theme.layoutSpacing['spacing/400']} + 4px)`,
    height: `calc( ${theme.layoutSpacing['spacing/400']} + 4px)`,
    borderRadius: `calc(${theme.borderRadius['borderRadius/circle']} / 2)`,
  },
  unchecked: {
    border: `2px solid ${theme.color['border/onSurfaceCurrent/default']}`,
  },
  disabledUnchecked: {
    borderColor: theme.color['border/onSurfaceCurrent/disabled'],
    backgroundColor: theme.color['bg/disabled'],
  },
  checkedIcon: {
    boxShadow: `inset 0px 0px 0px 2px ${theme.color['bg/surfacePrimary/default']}, inset 0px 0px 0px 5px ${theme.color['bg/surface100/default']}, inset 0px 0px 0px 10px ${theme.color['bg/surfacePrimary/default']}`,
  },
  disabledCheckedIcon: {
    boxShadow: `inset 0px 0px 0px 2px ${theme.color['bg/disabled']}, inset 0px 0px 0px 5px ${theme.color['bg/surface100/default']}, inset 0px 0px 0px 10px ${theme.color['bg/disabled']}`,
  },
}));

export default useStyles;
