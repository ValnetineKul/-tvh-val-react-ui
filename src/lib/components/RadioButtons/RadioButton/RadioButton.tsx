import type { FC } from 'react';
import React from 'react';
import FormControlLabel from '@mui/material/FormControlLabel';
import { default as MuiRadio } from '@mui/material/Radio';

import Typography from '../../Typography';
import useStyles from './RadioButton.styles';
import type { DataAttributes } from '../../../types/common';

export interface RadioButtonProps {
  label: string;
  value: string;
  name?: string;
  checked?: boolean;
  disabled?: boolean;
  className?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  inputProps?: React.HTMLAttributes<HTMLInputElement> & DataAttributes;
}

const RadioButton: FC<RadioButtonProps> = ({
  label,
  value,
  name,
  checked,
  disabled = false,
  className,
  onChange,
  inputProps,
}) => {
  const { classes, cx } = useStyles();

  const renderIcon = () => {
    return (
      <span
        className={cx(classes.icon, {
          [classes.disabledUnchecked]: disabled,
          [classes.unchecked]: !disabled,
        })}
      />
    );
  };

  const renderCheckedIcon = () => {
    return (
      <span
        className={cx(classes.icon, {
          [classes.disabledCheckedIcon]: disabled,
          [classes.checkedIcon]: !disabled,
        })}
      />
    );
  };

  const renderRadio = () => {
    return (
      <MuiRadio
        inputProps={inputProps}
        name={name}
        value={value}
        className={classes.radio}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
        icon={renderIcon()}
        checkedIcon={renderCheckedIcon()}
        disableRipple
      />
    );
  };

  if (!label) {
    return renderRadio();
  }

  return (
    <FormControlLabel
      classes={{ root: classes.root, label: cx(classes.label, className), disabled: classes.disabled }}
      label={<Typography component="span">{label}</Typography>}
      value={value}
      control={renderRadio()}
    />
  );
};

export default RadioButton;
