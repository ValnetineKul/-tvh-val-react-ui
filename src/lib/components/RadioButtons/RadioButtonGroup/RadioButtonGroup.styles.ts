import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'RadioButtonGroup' })((theme) => ({
  root: {
    '& label': {
      marginBottom: theme.layoutSpacing['spacing/300'],
      '&:last-child': {
        marginBottom: theme.layoutSpacing['spacing/000'],
      },
    },
  },
  label: {
    display: 'inline-block',
    color: theme.color['text/onSurfaceCurrent/default'],
    marginBottom: theme.layoutSpacing['spacing/300'],
  },
  asterisk: {
    color: theme.color['text/onSurfaceCurrent/error'],
  },
}));

export default useStyles;
