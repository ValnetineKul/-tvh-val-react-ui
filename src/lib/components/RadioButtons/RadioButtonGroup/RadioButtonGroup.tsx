import type { FC, ReactNode } from 'react';
import React from 'react';
import FormLabel from '@mui/material/FormLabel';
import RadioGroup from '@mui/material/RadioGroup';
import FormControl from '@mui/material/FormControl';
import Typography from '../../Typography';
import useStyles from './RadioButtonGroup.styles';

export interface RadioButtonGroupProps {
  children: ReactNode;
  value: string;
  label?: string;
  required?: boolean;
  name?: string;
  className?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>, value: string) => void;
}

const RadioButtonGroup: FC<RadioButtonGroupProps> = ({
  children,
  value,
  label,
  required = false,
  name,
  className,
  onChange,
}) => {
  const { classes } = useStyles();

  return (
    <FormControl component="fieldset" className={className}>
      {label && (
        <FormLabel component="legend">
          <Typography variant="body400" component="span" className={classes.label}>
            {label}
            {required && <span className={classes.asterisk}>&nbsp;*</span>}
          </Typography>
        </FormLabel>
      )}
      <RadioGroup value={value} onChange={onChange} classes={{ root: classes.root }} name={name}>
        {children}
      </RadioGroup>
    </FormControl>
  );
};

export default RadioButtonGroup;
