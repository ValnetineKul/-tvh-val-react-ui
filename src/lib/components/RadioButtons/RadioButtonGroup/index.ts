import RadioButtonGroup from './RadioButtonGroup';

export type { RadioButtonGroupProps } from './RadioButtonGroup';
export default RadioButtonGroup;
