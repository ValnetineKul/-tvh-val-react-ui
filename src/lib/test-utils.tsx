import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
