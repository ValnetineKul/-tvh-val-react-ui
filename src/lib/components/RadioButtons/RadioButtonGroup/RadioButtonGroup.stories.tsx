import type { ComponentProps } from 'react';
import React, { useState } from 'react';
import { createMeta } from '../../../story-utils';
import type { StoryTemplate } from '../../../story-utils';
import RadioButton from '../RadioButton';
import RadioButtonGroup from './RadioButtonGroup';

type Props = ComponentProps<typeof RadioButtonGroup>;
type Args = Omit<Props, 'value' | 'onChange' | 'children'>;

const Template = ((args: Args) => {
  const [value, setValue] = useState('option1');

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setValue((event.target as HTMLInputElement).value);
  };

  return (
    <RadioButtonGroup {...args} value={value} onChange={handleChange}>
      <RadioButton value="option1" label="RadioButton option" />
      <RadioButton value="option2" label="RadioButton option" />
      <RadioButton value="option3" label="RadioButton option" />
      <RadioButton value="option4" label="RadioButton option" />
      <RadioButton value="disabled" disabled label="Disabled option" />
    </RadioButtonGroup>
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  label: 'Label',
  required: true,
  name: 'radio-buttons',
};
Primitive.storyName = 'RadioButtonGroup';

export default createMeta({
  component: RadioButtonGroup,
  title: 'Components/RadioButtons/RadioButtonGroup',
  argTypes: {
    onChange: {
      table: {
        type: {
          detail:
            'Callback fired when a radio button is selected: (event: React.ChangeEvent<HTMLInputElement>) => void.',
        },
      },
    },
    value: {
      control: {
        disable: true,
      },
      table: {
        type: {
          detail: 'Value of the selected radio button.',
        },
      },
    },
  },
});
