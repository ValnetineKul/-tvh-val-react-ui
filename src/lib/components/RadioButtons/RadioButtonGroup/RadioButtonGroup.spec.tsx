import React from 'react';
import { render, screen } from '../../../test-utils';
import RadioButton from '../RadioButton';
import RadioButtonGroup from './RadioButtonGroup';

describe('RadioButtonGroup', () => {
  it('Should render correctly', () => {
    const onChange = jest.fn();
    const { container } = render(
      <RadioButtonGroup value="value" onChange={onChange}>
        Children
      </RadioButtonGroup>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should show label', () => {
    const onChange = jest.fn();
    render(
      <RadioButtonGroup value="value" onChange={onChange} label="label">
        Children
      </RadioButtonGroup>
    );
    expect(screen.getByText('label')).toBeInTheDocument();
  });

  it('Should show label with *', () => {
    const onChange = jest.fn();
    render(
      <RadioButtonGroup value="value" onChange={onChange} label="label" required>
        Children
      </RadioButtonGroup>
    );
    expect(screen.getByText('*')).toBeInTheDocument();
  });

  it('Should set "name" attribute in inputs with type="radio" to reference the value of the control', () => {
    render(
      <RadioButtonGroup value="value" onChange={() => null} name="radio-buttons">
        <RadioButton value="option" label="RadioButton option" />
      </RadioButtonGroup>
    );
    const radioButton = screen.getByRole('radio') as HTMLInputElement;
    expect(radioButton).toHaveAttribute('name', 'radio-buttons');
  });
});
