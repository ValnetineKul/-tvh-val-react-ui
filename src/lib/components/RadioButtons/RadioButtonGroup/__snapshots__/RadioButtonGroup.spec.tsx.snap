// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup Should render correctly 1`] = `
<fieldset
  class="MuiFormControl-root mui-1nrlq1o-MuiFormControl-root"
>
  <div
    class="tss-1gvbub8-RadioButtonGroup-root MuiFormGroup-root mui-dmmspl-MuiFormGroup-root"
    role="radiogroup"
  >
    Children
  </div>
</fieldset>
`;
